hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieFileGroupReader.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.hudi.common.config.HoodieReaderConfig;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.model.HoodieBaseFile;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordMerger;
+import org.apache.hudi.common.model.HoodieTableQueryType;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.log.HoodieMergedLogRecordReader;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.ClosableIterator;
+import org.apache.hudi.common.util.collection.EmptyIterator;
+import org.apache.hudi.common.util.collection.Pair;
+
+import org.apache.avro.Schema;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.hudi.common.fs.FSUtils.getRelativePartitionPath;
+import static org.apache.hudi.common.table.HoodieTableConfig.RECORD_MERGER_STRATEGY;
+import static org.apache.hudi.common.util.ConfigUtils.getBooleanWithAltKeys;
+import static org.apache.hudi.common.util.ConfigUtils.getIntWithAltKeys;
+import static org.apache.hudi.common.util.ConfigUtils.getStringWithAltKeys;
+
+/**
+ * A file group reader that iterates through the records in a single file group.
+ * <p>
+ * This should be used by the every engine integration, by plugging in a
+ * {@link HoodieReaderContext<T>} implementation.
+ *
+ * @param <T> The type of engine-specific record representation, e.g.,{@code InternalRow}
+ *            in Spark and {@code RowData} in Flink.
+ */
+public final class HoodieFileGroupReader<T> implements Closeable {
+  private final HoodieReaderContext<T> readerContext;
+  private final Option<HoodieBaseFile> baseFilePath;
+  private final Option<List<String>> logFilePathList;
+  private final Configuration hadoopConf;
+  private final TypedProperties props;
+  // Byte offset to start reading from the base file
+  private final long start;
+  // Length of bytes to read from the base file
+  private final long length;
+  // Key to record and metadata mapping from log files
+  private final Map<String, Pair<Option<T>, Map<String, Object>>> logFileRecordMapping = new HashMap<>();
+  private final FileGroupReaderState readerState = new FileGroupReaderState();
+  private ClosableIterator<T> baseFileIterator;
+  // This is only initialized and used after all records from the base file are iterated
+  private Iterator<Pair<Option<T>, Map<String, Object>>> logRecordIterator;
+  private HoodieRecordMerger recordMerger;
+
+  T nextRecord;
+
+  public HoodieFileGroupReader(HoodieReaderContext<T> readerContext,
```

Comment:
I don't see any opportunity for pushed down filters? Have we considered that for the design?